src/main.cpp

```diff
@@ -4070,6 +4073,53 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
     return true;
 }
 
+bool RewindBlockIndex(const Consensus::Params& params)
+{
+    LOCK(cs_main);
+
+    int nHeight = 1;
+    while (nHeight <= chainActive.Height()) {
+        if (IsWitnessEnabled(chainActive[nHeight - 1], params) && !(chainActive[nHeight]->nStatus & BLOCK_OPT_WITNESS)) {
+            break;
+        }
+        nHeight++;
+    }
+
+    // nHeight is now the height of the first insufficiently-validated block, or tipheight + 1
+    CValidationState state;
+    CBlockIndex* pindex = chainActive.Tip();
+    while (chainActive.Height() >= nHeight) {
+        if (!DisconnectTip(state, Params().GetConsensus(), true)) {
```

Comment:
To better support pruning nodes that might upgrade post-activation, I suggest we add a check  before this call to `DisconnectTip()`, something like:
```
if (fHavePruned && !(chainActive.Tip() & BLOCK_HAVE_DATA)) {
    // graceful error telling user to reindex and redownload the whole block chain to upgrade post-segwit activation
}
```